(PE-23319) add a network check to beaker-pe

mwbutcher opened the pull request:
Prior to this change, we would see potential
network related transients without data on
network health to back it up.

This commit adds a pre-test check to determine
if we can get to required network resources.
If this test fails, the test will continue after
printing a warning that includes the hostname
where the check failed and the target that
couldn't be reached.

Please delete any headings that don't apply to this Pull Request (PR).

#### What's this PR do?
#### Who would you like to review this PR?

@puppetlabs/integration, @puppetlabs/beaker (repo owners)

#### Should any of this be tested outside the normal PR CI cycle?
#### Any background context you want to provide?
#### Questions for reviewers?

barleyj-puppet commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -448,6 +455,29 @@ def deploy_frictionless_to_master(host)
+          logger.notify("Checking the availability of network resources.")
+          hosts.each do |host|
+            network_resources.each do |resource|
+              on host, "curl -s #{resource} > /dev/null", :acceptable_exit_codes => [0,6]
```
Should we add 35 to the exit codes?

mwbutcher replied:
yeah, in fact we should probably add a wildcard since this should be non-blocking.

barleyj-puppet commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -448,6 +455,29 @@ def deploy_frictionless_to_master(host)
+        #
+        def verify_network_resources(hosts, network_resources = NETWORK_RESOURCES)
+          logger.notify("Checking the availability of network resources.")
+          hosts.each do |host|
```
Should we also check that each host is available to each host? This would check for conditions like network segmentation or vmpooler reaping a vm we're using.

mwbutcher replied:
Oh, that's a great idea!

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/277/

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/278/

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/279/

barleyj-puppet commented:
👍  

I'll merge later this morning to give others time to review if they want.

ericwilliamson commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -448,6 +455,35 @@ def deploy_frictionless_to_master(host)
+          logger.notify("Checking the availability of network resources.")
+          hosts.each do |host|
+            network_resources.each do |resource|
+              on host, "curl -s #{resource} > /dev/null", :acceptable_exit_codes => (0...256)
```
instead of `:acceptable_exit_codes => (0...256)`

you can do

```
:accept_all_exit_codes => true
```


https://github.com/puppetlabs/beaker/blob/696b223898cf6daf836c33b88e6355d3777238f8/lib/beaker/dsl/helpers/host_helpers.rb#L15

highb replied:
Alternatively, we could be more specific here, since ping should only return exit codes 0, 1, and 2. I think if it returns outside that range, something has gone horribly wrong.

ericwilliamson commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -448,6 +455,35 @@ def deploy_frictionless_to_master(host)
+              end
+            end
+            hosts.each do |target_host|
+              on host, "ping -c1 #{target_host.host_hash[:vmhostname]} > /dev/null", :acceptable_exit_codes => (0...256)
```
Does this command work on all platforms?

`-c` doesn't seem to be a valid windows flag: https://technet.microsoft.com/en-us/library/cc940091.aspx

highb replied:
@mwbutcher Yeah you probably want `-n 1` here on windows.

highb replied:
Also, wouldn't we want the test to fail if it cannot ping the resources? So `0` should be the only valid exit code?

highb replied:
Or maybe build up an array of all non-zero exit codes and fail the dest if there are any? Seems silly to do this check and then proceed with testing the node if we can't reach critical resources.

ericwilliamson replied:
There needs to be a way to disable this, or partition the list of addresses - remember this is a public gem that people outside the company use. For released versions, customers don't need (and won't have access) too:


```
"enterprise.delivery.puppetlabs.net",
 "module-staging.puppet.com",
 "forge-aio01-petest.puppetlabs.com",
```

jpartlow replied:
Yes, one option might be that instead of having a constant array of resources to test, that the test suite (such as pe_acceptance_tests) provide that list.  It could insert into the options or into each host.  That way it does nothing outside of puppet, and each test suite can decide what resources it considers important.

jpartlow replied:
As far as the test failing if we can't find the resource; I'm not certain.  This is meant to be diagnostic, so I kind of feel that it's more important to get the information and any later test failures as well for comparison.

highb commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -38,6 +38,13 @@ module PEUtils
+            "aws.amazon.com",
+            "enterprise.delivery.puppetlabs.net",
+            "module-staging.puppet.com",
+            "forge-aio01-petest.puppetlabs.com",
```
This list should probably include the internal gem mirror, as well?

mwbutcher commented:
- I've added a check for windows that will result in a '-n' flag instead of '-c'
- The 'network_resources' array is now informed by options[:net_diag_hosts].  If that option isn't set, we'll just skip that check
- @jpartlow, you're right, this is just meant to be a diagnostic tool at this point.  We might want to fail fast on network issues in the future, but I think we should see how useful this is first.
- @highb regarding your ping vs curl point, a couple of those hosts don't respond to pings unfortunately
- @highb i've added rubygems.delivery.puppetlabs.net, but that host might actually be problematically slow.  We might need to get that fixed before we have tests waiting on it to respond.

jpartlow commented:
@mwbutcher I think Artifactory is handling gems internally now; I'm not sure if rubygems.delivery.puppetlabs.net redirects to that, or is obsolete, but I would check with releng if you haven't.

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/280/

mwbutcher commented:
@jpartlow thanks, just swapped rubygems.delivery.puppetlabs.net for artifactory.delivery.puppetlabs.net

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/281/

highb commented:
👍  I much prefer the more generic and re-usable approach here.

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/282/

highb commented:
Jenkins, test this please

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/283/

jpartlow commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -448,6 +448,41 @@ def deploy_frictionless_to_master(host)
+        # @return nil
+        #
+        # @api private
+        def verify_network_resources(hosts, network_resources = options[:net_diag_hosts])
```
I recommend one change -- rather than pulling net_diag_hosts out of options in the defaults, and forcing this method to know about the options state, instead how about requiring network_resources, and let do_install() fish out options[:net_diag_hosts] and explicitly pass it. The do_install() method already has options locally.

mwbutcher replied:
that makes sense

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/285/

jpartlow commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -448,6 +448,41 @@ def deploy_frictionless_to_master(host)
 
+        # Check for availability of required network resources
+        # @param [Array<Host>] hosts
+        # @param [Array<String>] network_resources, defaults to the value of options[:net_diag_hosts]
```
param doc is out of date (no longer has a default).  Other than that I'm :+1:

mwbutcher replied:
heh, good point.

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/286/
